### ISSUE TITLE
Logical client incompatible with KV Secrets Engine v2

### ISSUE THREAD

**Issue by larsp** [BetterCloud/vault-java-driver#114]
Hello!

Maybe I am totally off but it seems that the API is incompatible to the new KV secrets engine?

The new KV v2 write API is nesting secret key/values under the `data` field such as:
```
{
  "data": {
      "foo": "bar",
      "zip": "zap"
    }
}
```

Given the fact that `write` does not support nested structures and just calls `toString()` on any unknown type it seems that one can not use the API to write to the new KV engine?

Using a JSON string is not working since the string will be escaped which will lead to the following error:
```
{
  "errors": [
    "error converting input {\"foo\":\"bar\",\"zip\":\"zap\"} for field \"data\": '' expected a map, got 'string'"
  ]
}
```

Thanks

**Comment by larsp** [comment]
gentle ping @steve-perkins

**Comment by steve-perkins** [comment]
Sorry for the delays.  My company is just starting to look into the `0.10.x` upgrade, and so I'm a bit behind in my experience with the KV v2 write API.

Essentially, is the list of necessary changes below accurate and complete?  At least with regard to reads and writes with the logical API?

# (1) Different read and write paths

It looks like the KV v2 API expects a `data` second-level subpath on _writes_, which is omitted on _reads_.  For example, you POST some key-value data to:

`http://127.0.0.1:8200/v1/secret/data/my-secret`,
`http://127.0.0.1:8200/v1/secret/data/my-secret/my-subpath`, etc

... but GET the values from:

`http://127.0.0.1:8200/v1/secret/my-secret`,
`http://127.0.0.1:8200/v1/secret/my-secret/my-subpath`, etc

If I'm correct, then I'm not sure whether it makes more sense for the Java API to expect consistent path strings, and automatically insert the `data` subpath on writes?  Or just keep the Java API thin and dumb, and make it clear in the Javadocs that the user is responsible for using the correct paths on both reads and writes?

# (2) Handling of the data payload

With the KV v2 API it looks like the JSON key-value pairs being passed in the POST call need to be nested beneath a `data` object.  

Currently, the Java client using the old write API builds a JSON object with those key-value pairs as top-level fields.  Which is now incompatible.

So basically, I just need to nest the key-value pairs beneath a `data` object on write operations?  I believe I should already be covered for read operations, as both the old API and the v2 version both return the key-value pairs from a `data` object on GET's.

**Comment by delfuego** [comment]
I'm not entirely sure if the reads are still correct with the non-`data`-containing path; I haven't debugged further than knowing that my app, which reads data fine with a v1 K/V secret store, fails to read data with a v2 K/V secret store. I might be able to find time to dig into this a bit more over the next week, but that's what I discovered — and if your theory in (1) above were correct, then reads would have continued to succeed.

**Comment by steve-perkins** [comment]
@delfuego:  Thanks a lot.  It looks like I got my wires crossed when doodling with it this morning.  You are correct... it seems that the `data/` string needs to be present as a second-level subpath on both writes and reads.  I'll update my comment above to avoid confusion for future people reading it.

The overall question there still stands, though... would it be a good idea to automatically "inject" that `data` subpath in the paths provided?  

If it turns out that path strings ARE consistent for both reads and writes, then I'm now more inclined to keep the Java API agnostic, and simply depend on users to supply the correct paths that they really mean.  However, internally within my own company I may have to consider cutting a private release that does such path-fixing... as a migration step until we can get ALL our secret paths changed everywhere in code.

**Comment by mp911de** [comment]
Two more aspects come here into play: 

1. The Key/Value backend can be operated in both versions with different mount paths. 
2. Someone has to decide over the version that is in use. Right now, `vault-cli` issues a preflight request upon each Key/Value backend interaction to an internal Vault endpoint to determine the version. Another approach could be that the application code using the client knows which key-value API version to use.

**Comment by steve-perkins** [comment]
Thanks so much, @mp911de.  I took a look at your work on the Spring client, and was relieved to see that it's basically the same approach I was going with in my private proof-of-concept branch.  That is, inserting a `data` or `metadata` segment in mid-path, depending on the operation.  

I'm my POC, I'm just doing this as the standard, as I believe we will be moving to the new backend exclusively.  However, I agree that if both backend versions can be mounted simultaneously, then the final public driver will need the ability to support both.  I'm leaning toward making this a fixed config value, even if it means that developers might need to instantiate two driver instances, rather than taking the performance hit of doing a pre-flight check on every call.

One other thing, though.  In my testing, I've had to slightly modify the request payload for writes, and the expected response payload for reads.  In both cases, the key-value pairs being sent or received are now expected to be nested within an additional `data` object level.

For write operations, I see how you are accounting for this difference in your `VaultKeyValue1Template.put(...)` and `VaultKeyValue2Template.put(...)` implementations.  However, maybe it's just too early in the morning local time (!), but I'm not seeing how you address it for reads in the two `get(...)` implementations.  

Since the `vaultResponse.setData(data)` call is just populating a `Map<String, Object>` object... are you simply passing the outermost-data object here, and making it the caller's responsibility to know that this will contain a nested data object when using the new backend?

**Comment by mp911de** [comment]
Hey Steve, 

the class hierarchy might obfuscate some of the code paths. We introduced `VaultKeyValueAccessor` as abstract base class requiring `getJsonNode(…)` to be implemented in the subclasses:

* [`VaultKeyValue1Template`](https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/main/java/org/springframework/vault/core/VaultKeyValue1Template.java#L116)
* [`VaultKeyValue2Accessor`](https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/main/java/org/springframework/vault/core/VaultKeyValue2Accessor.java#L78) (common base class for `VaultKeyValue2Template` and `VaultVersionedKeyValueTemplate `)

Everything else happens in [`VaultKeyValueAccessor`](https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/main/java/org/springframework/vault/core/VaultKeyValueAccessor.java#L104).

We decided to provide a generic Key-Value API and a Versioned Key Value API to account for use cases where you just want to interact with secrets without touching aspects of versioning (use either one of the backend versions) and cases you want to interact additionally with version metadata.

**Comment by buttermilk-crypto** [comment]
Hi Steve,

The support for kv v2 is needed for my current project. Any chance of making your branch available? I understand it's work in progress. 

The kv v2 API seems quite different. It has new concepts and Hashicorp has new CLI - the kv command - to support this change. So they are not attempting to abstract away this difference.

**Comment by emailtomanishkr** [comment]
Hi,

We need support for versioned KV secret engine v2. Is there any date released when the vault-java-driver will support versioned KV secret engine?

**Comment by JuergenKindler** [comment]
Also would like to see support for v2 of KV secret engine.

**Comment by clatko** [comment]
Not a comprehensive port, but serves my purposes well.

https://github.com/BetterCloud/vault-java-driver/pull/144

**Comment by danybenjamin-tc** [comment]
When can we expect a release with this fix? The java driver remains useless to us as we're on KV V2

**Comment by jedidiahB** [comment]
@danybenjamin-tc #151 includes KV version 2 support, and was introduced yesterday.

**Comment by danybenjamin-tc** [comment]
Oh wow - that's great. Will we see a new release on mvn repo?

**Comment by steve-perkins** [comment]
Just published a version 4.0.0 build to Maven Central.  This will probably take several hours to show up in their indexes.

**Comment by jhoblitt** [comment]
Is this considered resolved or are there known issues with v2 support?

**Comment by steve-perkins-bc** [comment]
No (known) issues.  I've been going through open PR's over the past couple of days, but need to make a sweep through the issues board as well to clean things up.  Marking as resolved, anyone feel free to create a new issue for anything that arises.
